packages/report-server/src/__tests__/reportBuilder/transform/insert.test.ts

```diff
@@ -0,0 +1,214 @@
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { dataElementValue: 4 },
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { dataElementValue: 2 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 5 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+    ]);
+  });
+
+  it('can insert new rows at the beginning of the list', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+        position: 'start',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { dataElementValue: 4 },
+      { dataElementValue: 2 },
+      { dataElementValue: 5 },
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+    ]);
+  });
+
+  it('can insert specific new rows using a where clause', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+        position: 'after',
+        where: "not(eq($row.period, '20200101'))",
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+      { dataElementValue: 5 },
+    ]);
+  });
+
+  // USE CASES
+  it('can sum all values into a new totals row', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        where: 'eq($index, length($table))',
+        "'Total'": 'sum($all.value)',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+      { Total: 11 }
+    ]);
+  });
+
+  it('compare adjacent rows to insert between', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        where: 'and(exists($next.organisationUnit), not(eq($row.organisationUnit, $next.organisationUnit)))',
+        "'Total_BCD1'": "sum($allPrevious.BCD1)",
+        "'Total_BCD2'": "sum($allPrevious.BCD2)",
+      },
+      {
+        transform: 'insert',
+        where: 'eq($index, length($table))',
+        "'Total_BCD1'": "sum($all.BCD1) - sum($allPrevious.Total_BCD1)",
+        "'Total_BCD2'": "sum($all.BCD2) - sum($allPrevious.Total_BCD2)",
+      }
```

Comment:
Nice work on the `$next`! I think that will come in handy, and be a thing that "just works" when someone expects that if there's a previous, there's a next.

As a more general transform step that _should_ work across any number of categories, could we try out the following?
```suggestion
      {
        transform: 'insert',
        where: 'not(eq($row.organisationUnit, $next.organisationUnit))',
        "'Total_BCD1'": "sum($where(f($otherRow) = equalText($otherRow.organisationUnitCode, $row.organisationUnitCode)).BCD1)",
        "'Total_BCD2'": "sum($where(f($otherRow) = equalText($otherRow.organisationUnitCode, $row.organisationUnitCode)).BCD2)"
      }
```

Reply:
Yep that works perfect!